### ISSUE TITLE
julia-repl-mode fails oddly in buffers with "false" buffer-file-name

### ISSUE THREAD

**Issue by non-Jedi** [tpapp/julia-repl#41]
See discussion at syl20bnr/spacemacs#11314.

I think this bit of code: https://github.com/tpapp/julia-repl/blob/06678ed0cb07807f6cb153c6b0997edc6a18f22c/julia-repl.el#L557 doesn't properly account for buffers not associated with any file, but I could be barking up the wrong tree there since this is my first encounter with `when-let`, so it's taking me a bit to grok the implications.

Problem reproduces by invoking `julia-repl-mode` in a `*scratch*` buffer.

**Comment by tpapp** [comment]
I think that `when-let` should account precisely for the case when `(buffer-file-name)` is nil.

You can test (`M-:` eval) either
```emacs-lisp
(when-let ((filename (buffer-file-name)))
    (setq-local default-directory (file-name-directory filename)))
```
or simply
```emacs-lisp
(buffer-file-name)
```
in the relevant buffer to see if something funny is going on.

**Comment by non-Jedi** [comment]
Hi @tpapp. Finally circling back around to this. The problem is that the buffers created by `C-c '` in org-mode have `buffer-file-name` set to e.g. `"/home/adam/org/solvay/log.org[*Org Src log.org[ jupyter-julia ]*]"` for `#+BEGIN_SRC` blocks in `/home/adam/org/solvay/log.org`.

This isn't just a peculiarity of ob-ipython but appears to be how all `#+BEGIN_SRC` blocks work in org-mode. For example, `buffer-file-name` for a python block in *scratch* buffer is `"[*Org Src *scratch*[ python ]*]"`.

**Comment by tpapp** [comment]
@non-Jedi: suggestions/PRs to fix this are welcome.

**Comment by non-Jedi** [comment]
I'll see if I can take a deeper look this evening, but from scrolling through the emacs-lisp docs, `buffer-file-truename` looks like it doesn't return the weird fake filename in the buffers in question.

**Comment by non-Jedi** [comment]
Background: flycheck ran into this problem in March (https://lists.gnu.org/archive/html/emacs-orgmode/2018-03/msg00226.html). Looks like org-mode should have changed to no longer behave this way, but I'm still seeing it, so I wonder if it just hasn't made it into a release yet (https://lists.gnu.org/archive/html/emacs-orgmode/2018-04/msg00486.html).

**Comment by non-Jedi** [comment]
Should be fixed in org-mode release 9.2 although I am no longer using julia-repl, so I haven't personally tested this.

**Comment by tpapp** [comment]
Thanks for following up on this, if anyone experiences this issue please feel free to reopen. I can't replicate on 9.2.
